### PR TITLE
[Fix] CombineTensorPatches

### LINKED ISSUE #1557: Incorrect shape while using CombineTensorPatches
### Describe the bug

Currently `CombineTensorPatches` does not accurately invert the results of `ExtractTensorPatches`. This can be observed by running the following script.

### Reproduction steps

```bash
import torch

from kornia.contrib import CombineTensorPatches, ExtractTensorPatches

# Original size
h, w = 16, 16
# Window size
win = 2

image = torch.randn(2, 3, h, w)
print(image.shape)
tiler = ExtractTensorPatches(window_size=win, stride=win, padding=0)
merger = CombineTensorPatches(window_size=win, unpadding=0)
image_tiles = tiler(image)
new_image = merger(image_tiles)
print(new_image.shape)
assert (image == new_image).all()
```

Instead of returning the original shape (2,3,16,16), `CombineTensorPatches` will return (32, 3, 4, 4)

### Expected behavior

image and new_image should have the same shape - (2,  3, 16, 16)

### Environment

```shell
I used the devcontainer


PyTorch version: 1.10.2+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 11 (bullseye) (x86_64)
GCC version: (Debian 10.2.1-6) 10.2.1 20210110
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.8.12 (default, Dec 21 2021, 10:45:09)  [GCC 10.2.1 20210110] (64-bit runtime)
Python platform: Linux-5.4.0-96-generic-x86_64-with-glibc2.2.5
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy==0.931
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.2
[pip3] pytest-mypy==0.8.1
[pip3] torch==1.10.2
[pip3] torchvision==0.11.3
[conda] Could not collect


```
### Additional context

This error is caused by `CombineTensorPatches` not taking into account the original size of the image. It can be solved by setting `window_size = (original_size[0] // window_size[0], original_size[1] // window_size[1])`. I've opened a PR to implement this fix.

### PR DESCRIPTION
**Pull request by ashnair1**
Fixes #1557 

#### Type of change

- [x] 🧪 Tests Cases
- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

**Comment by ashnair1** [comment]
@edgarriba  @shijianjian  Any thoughts on this PR?

File: kornia/contrib/extract_patches.py [hunk 1]
```diff
@@ -107,7 +107,8 @@ class CombineTensorPatches(nn.Module):
 
     Args:
         patches: patched tensor.
-        window_size: the size of the sliding window and the output patch size.
+        orig_size: the size of the original tensor and the output patch size.
```
Comment (edgarriba) [inline kornia/contrib/extract_patches.py:110]:
```suggestion
        original_size: the size of the original tensor and the output patch size.
```

**Review by edgarriba — APPROVED** [review]
LGTM. Just comment on better variable naming. @shijianjian review and merge

**Comment by ashnair1** [comment]
Just found that this doesn't fully fix the problem. With non-zero padding the error still appears. I've added a test to account for this but I'll need to look into this a bit more to full solve the issue.

**Comment by ashnair1** [comment]
Ok, think I've got it now. With the new fix patching and merging works as expected even for rectangular images. Only point to note is that the original image dimensions should be even for unpadding to work correctly. Additionally the padded image must be divisible by the window size.

**Comment by ashnair1** [comment]
Just noticed that this fix could also be ported to PatchSequential's [restore_from_patches](https://github.com/kornia/kornia/blob/a598482ffe078b2ab68f1ad74f4874734d557f7d/kornia/augmentation/container/patch.py#L220). Or we could just use the `combine_tensor_patches` instead.

**Comment by ashnair1** [comment]
Pinging @edgarriba @shijianjian

File: kornia/contrib/extract_patches.py [hunk 2]
```diff
@@ -125,20 +126,25 @@ class CombineTensorPatches(nn.Module):
...
         super().__init__()
+        self.orig_size: Tuple[int, int] = _pair(original_size)
         self.window_size: Tuple[int, int] = _pair(window_size)
         pad: Tuple[int, int] = _pair(unpadding)
         self.unpadding: Tuple[int, int, int, int] = (pad[0], pad[0], pad[1], pad[1])
 
     def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
-        return combine_tensor_patches(input, self.window_size, stride=self.window_size, unpadding=self.unpadding)
+        return combine_tensor_patches(
+            input, self.orig_size, self.window_size, stride=self.window_size, unpadding=self.unpadding
```
Comment (edgarriba) [inline kornia/contrib/extract_patches.py:141]:
```suggestion
            input, self.original_size, self.window_size, stride=self.window_size, unpadding=self.unpadding
```

**Review by edgarriba — APPROVED** [review]
LGTM. @shijianjian merge once we have green lights in the CI

File: kornia/contrib/extract_patches.py [hunk 3]
```diff
@@ -125,20 +126,25 @@ class CombineTensorPatches(nn.Module):
...
+        )
 
 
 def combine_tensor_patches(
     patches: torch.Tensor,
+    orig_size: Tuple[int, int] = (16, 16),
```
[the remark this answers is not in the document]
Comment (ashnair1) [reply kornia/contrib/extract_patches.py]: Since the function wasn't working correctly before, is there a need to maintain backwards compatibility?

[the remark this answers is not in the document]
Comment (ashnair1) [reply kornia/contrib/extract_patches.py]: While on the topic, wouldn't it better to remove the default arguments for combine and extract?  While using these a user should explicitly designate the window_size, stride and original_size (for combine) parameters so that they get a reasonable result. Otherwise we run the risk of them specifying one parameter and getting an incorrect result or an OOM since default stride is 1.

File: kornia/contrib/extract_patches.py [hunk 4]
```diff
@@ -125,20 +126,25 @@ class CombineTensorPatches(nn.Module):
 
     def __init__(
         self,
+        original_size: Union[int, Tuple[int, int]],
         window_size: Union[int, Tuple[int, int]],
         unpadding: Union[int, Tuple[int, int]] = 0,
```
Comment (shijianjian) [inline kornia/contrib/extract_patches.py:131]: missing `stride`

File: kornia/contrib/extract_patches.py [hunk 5]
```diff
@@ -167,17 +174,28 @@ def combine_tensor_patches(
...
+        hpad_check = (original_size[0] + unpadding[0] + unpadding[1]) % window_size[0] == 0
+        wpad_check = (original_size[1] + unpadding[2] + unpadding[3]) % window_size[1] == 0
+
+        if not hpad_check or not wpad_check:
+            raise NotImplementedError("Insufficient padding")
+
         window_size = (
-            window_size[0] + (unpadding[0] + unpadding[1]) // window_size[0],
-            window_size[1] + (unpadding[2] + unpadding[3]) // window_size[1]
+            (original_size[0] + (unpadding[0] + unpadding[1])) // window_size[0],
```
Comment (shijianjian) [inline kornia/contrib/extract_patches.py:189]: I am wondering what is the definition of `window_size` here? why it changes so dramatically?`window_size==image_size`?

File: kornia/contrib/extract_patches.py (hunk 4 above)
Comment (ashnair1) [reply kornia/contrib/extract_patches.py:131]: `combine_tensor_patches` currently only supports `stride == window_size`. Refer [here](https://github.com/kornia/kornia/blob/62f114e7883292ae6bb38874cfaca2d98c13a160/kornia/contrib/extract_patches.py#L165).

File: kornia/contrib/extract_patches.py (hunk 5 above)
Comment (ashnair1) [reply kornia/contrib/extract_patches.py:189]: The window_size that is passed to `combine_tensor_patches` is the window_size used for `extract_tensor_patches`. 
window_size is not the same as image size.


Consider the following code snippet


```
import torch

from kornia.contrib import CombineTensorPatches, ExtractTensorPatches

h, w = 8, 8
win = 4
pad = 2

image = torch.randn(1, 1, h, w)
print(image.shape)
tiler = ExtractTensorPatches(window_size=win, stride=win, padding=pad)
merger = CombineTensorPatches(original_size=(h, w), window_size=win,  unpadding=pad)
image_tiles = tiler(image)
print(image_tiles.shape)
new_image = merger(image_tiles)
print(new_image.shape)
assert (image == new_image).all()
```


Here 
original_image_size = (8,8)
window_size used for extraction = (4, 4) 
padding = (2,2,2,2)

In combine_tensor_patches we compute window_size to take unpadding into account, so
```
combine_window_size = window_size = (
            (original_size[0] + (unpadding[0] + unpadding[1])) // window_size[0],
            (original_size[1] + (unpadding[2] + unpadding[3])) // window_size[1],
        )
```
which comes out to be (3, 3)

File: kornia/contrib/extract_patches.py (hunk 4 above)
Comment (edgarriba) [reply kornia/contrib/extract_patches.py:131]: Might be worth to add a note somewhere about this detail

File: kornia/contrib/extract_patches.py (hunk 5 above)
Comment (edgarriba) [reply kornia/contrib/extract_patches.py:189]: would nice to have in a posterior PR this snippet as an example with visualisation

Comment (ashnair1) [reply kornia/contrib/extract_patches.py:189]: Sure. I could do that in a follow up PR. Where do you recommend I add it?

File: kornia/contrib/extract_patches.py (hunk 4 above)
Comment (ashnair1) [reply kornia/contrib/extract_patches.py:131]: We could add the stride param and forward it to `combine_tensor_patches` where a `NotImplementedError` will be thrown.

File: kornia/contrib/extract_patches.py [hunk 6]
```diff
@@ -107,7 +107,8 @@ class CombineTensorPatches(nn.Module):
 
     Args:
         patches: patched tensor.
-        window_size: the size of the sliding window and the output patch size.
+        original_size: the size of the original tensor and the output patch size.
```
Comment (edgarriba) [inline kornia/contrib/extract_patches.py:110]: The docs above must be updated, it mentions padding, stride, etc. Also, worth to cross link a sentence saying that this method it’s recommended to be used after extract patches. See: https://github.com/kornia/kornia/blob/master/kornia/geometry/transform/imgwarp.py#L84

Comment (ashnair1) [reply kornia/contrib/extract_patches.py:110]: Docs have been updated.

File: kornia/contrib/extract_patches.py (hunk 5 above)
Comment (edgarriba) [reply kornia/contrib/extract_patches.py:196]: Great, https://github.com/kornia/tutorials/tree/master/source
